File: frontend/src/utils/__tests__/metadata.test.ts
import { describe, expect, it } from "vitest";

import { shouldShowMetadataMismatch } from "../metadata";

describe("shouldShowMetadataMismatch", () => {
  it("shows metadata only for mismatched file types", () => {
    expect(
      shouldShowMetadataMismatch({
        extensionType: "image/jpeg",
        detectedType: "image/webp",
        typeMismatch: true,
      })
    ).toBe(true);

    expect(
      shouldShowMetadataMismatch({
        extensionType: "image/jpeg",
        detectedType: "image/jpeg",
        typeMismatch: false,
      })
    ).toBe(false);
  });
});

File: files/metadata_test.go
package files

import (
	"strings"
	"testing"
)

func TestDetectMetadataReportsMimeMismatch(t *testing.T) {
	t.Parallel()

	const webpHeader = "RIFF\x1a\x00\x00\x00WEBPVP8 "

	metadata, err := DetectMetadata(".jpg", strings.NewReader(webpHeader))
	if err != nil {
		t.Fatalf("DetectMetadata returned error: %v", err)
	}

	if metadata.ExtensionType != "image/jpeg" {
		t.Fatalf("ExtensionType = %q, want image/jpeg", metadata.ExtensionType)
	}
	if metadata.DetectedType != "image/webp" {
		t.Fatalf("DetectedType = %q, want image/webp", metadata.DetectedType)
	}
	if !metadata.TypeMismatch {
		t.Fatal("TypeMismatch = false, want true")
	}
}

func TestDetectMetadataDoesNotReportMatchingMime(t *testing.T) {
	t.Parallel()

	const pngHeader = "\x89PNG\r\n\x1a\n\x00\x00\x00\rIHDR"

	metadata, err := DetectMetadata(".png", strings.NewReader(pngHeader))
	if err != nil {
		t.Fatalf("DetectMetadata returned error: %v", err)
	}

	if metadata.ExtensionType != "image/png" {
		t.Fatalf("ExtensionType = %q, want image/png", metadata.ExtensionType)
	}
	if metadata.DetectedType != "image/png" {
		t.Fatalf("DetectedType = %q, want image/png", metadata.DetectedType)
	}
	if metadata.TypeMismatch {
		t.Fatal("TypeMismatch = true, want false")
	}
}
